libzmix/bulletproofs_amcl/src/r1cs/gadgets/vsmt_4.rs

```diff
@@ -0,0 +1,613 @@
+use super::helper_constraints::constrain_lc_with_scalar;
+use crate::errors::R1CSError;
+use crate::r1cs::linear_combination::AllocatedQuantity;
+use crate::r1cs::{ConstraintSystem, LinearCombination, Prover, R1CSProof, Variable, Verifier};
+use amcl_wrapper::constants::{MODBYTES, NLEN};
+use amcl_wrapper::field_elem::FieldElement;
+use amcl_wrapper::group_elem::GroupElement;
+use amcl_wrapper::group_elem_g1::{G1Vector, G1};
+use merlin::Transcript;
+use std::time::{Duration, Instant};
+
+use rand::{CryptoRng, RngCore};
+
+use super::helper_constraints::poseidon::{
+    PoseidonParams, Poseidon_hash_4, Poseidon_hash_4_constraints, SboxType, PADDING_CONST,
+};
+use crate::r1cs::gadgets::poseidon::{allocate_statics_for_prover, allocate_statics_for_verifier};
+use std::collections::HashMap;
+
+type DBVal = [FieldElement; 4];
+type ProofNode = [FieldElement; 3];
+
+/// Depth of the tree.
+/// Has to be a multiple of 4.
+// TODO: Remove above restriction.
+pub const TreeDepth: usize = 16;
+
+/// Number of bytes to represent leaf index
+pub const LeafIndexBytes: usize = TreeDepth / 4;
+
+/// Get a base 4 representation of the given `scalar`. Only process `limit_bytes` of the scalar
+pub fn get_base_4_repr(scalar: &FieldElement, limit_bytes: usize) -> Vec<u8> {
+    if limit_bytes > MODBYTES {
+        panic!(
+            "limit_bytes cannot be more than {} but found {}",
+            MODBYTES, limit_bytes
+        )
+    }
+    let d = limit_bytes * 4; // number of base 4 digits
+    let mut s = scalar.to_bignum();
+    s.norm();
+
+    let mut base_4 = vec![];
+    while (base_4.len() != d) && (!s.iszilch()) {
+        base_4.push(s.lastbits(2) as u8);
+        s.fshr(2);
+    }
+    while base_4.len() != d {
+        base_4.push(0);
+    }
+
+    base_4.reverse();
+    base_4
+}
+
+// TODO: ABSTRACT HASH FUNCTION BETTER
+/// Sparse merkle tree with width 4, .i.e each node has 4 children.
+pub struct VanillaSparseMerkleTree_4<'a> {
```

Comment:
Could we use a better flavor?!?

Reply:
Yes, I was trying a better variation but could not get that to work. This is the [one](https://github.com/lovesh/bulletproofs-r1cs-gadgets/blob/master/src/gadget_osmt.rs).